### ISSUE TITLE
Information about tcn training

### ISSUE THREAD

**Issue by gabrielecaddeo** [facebookresearch/MidasTouch#3]
Hi all,
Thanks for sharing with us your great work.

I am trying to use your code with my DIGIT sensor. In the paper, you mentioned a training procedure for the TCN (Tactile Code Network)
![tcn_training](https://user-images.githubusercontent.com/49904924/209107202-47814d7e-3686-4d7c-b985-ed17827b4c50.png)

but I cannot see any training code for this part. I was wondering if you can kindly share that part of code or if I am missing something.

Thanks,
Gabriele

**Comment by suddhu** [comment]
Hi @gabrielecaddeo!

Thank you for the question! The point-cloud to embedding training is currently not open-sourced, it will take a while for me to organize this. I'll keep you updated when I find time to do this. A couple of points that may help you: 

- To recreate this yourself, you can refer to the [train.py](https://github.com/jac99/MinkLoc3D/blob/master/training/train.py) script in [MinkLoc3D](https://github.com/jac99/MinkLoc3D). We use the same training procedure, except with a dataloader for pointclouds generated by our [generate_data.py](https://github.com/facebookresearch/MidasTouch/blob/main/midastouch/data_gen/generate_data.py). Once we get the pointclouds we build postive-negative labels based on [this script](https://github.com/jac99/MinkLoc3D/blob/master/generating_queries/generate_training_tuples_baseline.py) in MinkLoc3D. I'm happy to answer any specific questions there! Another TCN alternative can be [PointNetVLAD](http://github.com/mikacuy/pointnetvlad/)
- As mentioned in the paper, we only fine-tune the pretrained weights of MinkLoc3D for tactile point clouds. So our weights file may directly work well for your application too. After all, only the TDN is sensor-dependent, the TCN doesn't depend on the DIGIT image characteristics. 

Thank you for your interest in the work, I'm happy to answer any specific implementation questions in this thread!

**Comment by gabrielecaddeo** [comment]
Hi @suddhu,

Thanks for your answer!



> Thank you for the question! The point-cloud to embedding training is currently not open-sourced, it will take a while for me to organize this. I'll keep you updated when I find time to do this. A couple of points that may help you:
> 
Okay, I will be waiting! Thanks
>     * To recreate this yourself, you can refer to the [train.py](https://github.com/jac99/MinkLoc3D/blob/master/training/train.py) script in [MinkLoc3D](https://github.com/jac99/MinkLoc3D). We use the same training procedure, except with a dataloader for pointclouds generated by our [generate_data.py](https://github.com/facebookresearch/MidasTouch/blob/main/midastouch/data_gen/generate_data.py). Once we get the pointclouds we build postive-negative labels based on [this script](https://github.com/jac99/MinkLoc3D/blob/master/generating_queries/generate_training_tuples_baseline.py) in MinkLoc3D. I'm happy to answer any specific questions there! Another TCN alternative can be [PointNetVLAD](https://github.com/mikacuy/pointnetvlad/)
> 

I will spend some time next week to implement this solution, thank you for the instructions!
>     * As mentioned in the paper, we only fine-tune the pretrained weights of MinkLoc3D for tactile point clouds. So our weights file may directly work well for your application too. After all, only the TDN is sensor-dependent, the TCN doesn't depend on the DIGIT image characteristics.
> 
I get your point. However, since the backgrounds of our DIGITs are different I wanted to see how much the performances would increase by fine-tuning the TCN with the proper backgrounds

> Thank you for your interest in the work, I'm happy to answer any specific implementation questions in this thread!
>
Thank you very much, we can keep this issue open for eventual future question!

**Comment by gabrielecaddeo** [comment]
> I get your point. However, since the backgrounds of our DIGITs are different I wanted to see how much the performances would increase by fine-tuning the TCN with the proper backgrounds
> 
Sorry, I got confused, my answer is wrong. Do you confirm me that the fine-tuning of the TCN depends only on the 3d patch obtained with TACTO and not on the rgb images?
Thank you very much for your time!

**Comment by suddhu** [comment]
Yup, the TCN solves a pointcloud retrieval problem, which is purely based on the 3D patches. The variations in background will affect the image to 3D part only, which is the TDN. The training scripts for the TDN from Tacto data are [here](https://github.com/facebookresearch/MidasTouch/blob/main/midastouch/contrib/tdn_fcrn/train.py).
